core/src/toga/dialogs.py

```diff
@@ -168,14 +188,31 @@ def __init__(
         )
 
 
-class OpenFileDialog(Dialog):
+class OpenFileDialog(Dialog[TDialogResult]):
+    @overload
+    def __init__(
+        self: OpenFileDialog[Path],
+        title: str,
+        initial_directory: Path | str | None = None,
+        file_types: list[str] | None = None,
+        multiple_select: Literal[False] = ...,
+    ) -> None: ...
+    @overload
+    def __init__(
+        self: OpenFileDialog[list[Path]],
+        title: str,
+        initial_directory: Path | str | None = None,
+        file_types: list[str] | None = None,
+        *,
+        multiple_select: Literal[True],
```

Comment:
This isn't exactly accurate (`multiple_select` isn't a keyword-only argument) but we'd need a myriad overloads otherwise, see https://github.com/python/typing/discussions/1326#discussioncomment-4568784.

Reply:
My bigger concern here is the [impact on documentation](https://toga--3267.org.readthedocs.build/en/3267/reference/api/resources/dialogs.html#toga.OpenFileDialog). Is the explicit enumeration of "all the values a boolean can have" as overrides strictly necessary? What's wrong with just `bool?

Reply:
The function signature in the docs is missing the return type, but the purpose is to vary the return type based on the value of `multiple_select`. When `multiple_select` is `True`, you get a `list[Path]`, and when it's `False`, you get a mere `Path` (see the type of `self` on line 194 and then on 202). We could just return `Path | list[Path]`, but that's probably worse for typing users 'cause now instead of an `Any` they can use freely, they'd have to check whether the return value is or isn't a list.

Reply:
Ah - I missed the difference in the narrowing of the `self` argument. That makes sense; thanks for clarifying.